File: backend/app/api/v1/endpoints/predictions.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List
from app.api.dependencies import get_db
from app.services.prediction_service import PredictionService
from app.schemas.prediction import PredictionOut

router = APIRouter()

@router.post("/students/{student_id}/predict", response_model=PredictionOut)
async def predict_student_risk(
    student_id: str,
    model_type: str = Query("LATE_STAGE", pattern="^(EARLY_WARNING|LATE_STAGE)$"),
    db: AsyncSession = Depends(get_db)
):
    return await PredictionService.predict_student_risk(db, student_id, model_type)

@router.get("/students/{student_id}/predictions", response_model=List[PredictionOut])
async def get_student_predictions(
    student_id: str,
    db: AsyncSession = Depends(get_db)
):
    return await PredictionService.get_student_predictions(db, student_id)

@router.get("/predictions/{prediction_id}", response_model=PredictionOut)
async def get_prediction(
    prediction_id: str,
    db: AsyncSession = Depends(get_db)
):
    return await PredictionService.get_prediction(db, prediction_id)

File: frontend/src/api/predictions.ts
import apiClient from './client';
import { Prediction, Recommendation } from '../types/prediction';

export const predictStudentRisk = async (
  studentId: string,
  modelType: 'EARLY_WARNING' | 'LATE_STAGE' = 'LATE_STAGE'
): Promise<Prediction> => {
  const response = await apiClient.post<Prediction>(`/students/${studentId}/predict?model_type=${modelType}`);
  return response.data;
};

export const getStudentPredictions = async (studentId: string): Promise<Prediction[]> => {
  const response = await apiClient.get<Prediction[]>(`/students/${studentId}/predictions`);
  return response.data;
};

export const getPrediction = async (predictionId: string): Promise<Prediction> => {
  const response = await apiClient.get<Prediction>(`/predictions/${predictionId}`);
  return response.data;
};

export const generateRecommendation = async (
  predictionId: string,
  teacherNotes?: string,
  documentId?: string
): Promise<Recommendation> => {
  const response = await apiClient.post<Recommendation>(`/predictions/${predictionId}/recommendation`, {
    teacher_notes: teacherNotes || null,
    document_id: documentId || null
  });
  return response.data;
};

export const getRecommendation = async (recommendationId: string): Promise<Recommendation> => {
  const response = await apiClient.get<Recommendation>(`/recommendations/${recommendationId}`);
  return response.data;
};
